cmd/swarm/swarm-smoke/upload_and_sync.go

```diff
@@ -65,58 +63,93 @@ func uploadAndSyncCmd(ctx *cli.Context, tuid string) error {
 	}
 
 	// trigger debug functionality on randomBytes
-	e := trackChunks(randomBytes[:])
+	e := trackChunks(randomBytes[:], true)
 	if e != nil {
 		log.Error(e.Error())
 	}
 
 	return err
 }
 
-func trackChunks(testData []byte) error {
+func trackChunks(testData []byte, submitMetrics bool) error {
 	addrs, err := getAllRefs(testData)
 	if err != nil {
+		log.Error("cannot get refs", "err", err.Error())
```

Comment:
This error is already passed to the caller of this function. Do we need to log it also, since it is to the caller to decide what to do with it and the error is already logged.

Reply:
If we don't log it here, sometimes it is difficult to know which RPC call actually failed. This is a problem with bigger files, when you reach size of frame size and various limitations.

Reply:
We can add more descriptive error messages where this function is called and error logged. There should not be the same error in two log lines.

Reply:
You do have the line number when this is logged, so you know which one actually emitted it.

Reply:
The point is that whenever `trackChunks` fails (which doesn't have an effect on the smoke tests, but has an effect on your debugging efforts), you want to know exactly where and why it failed while reviewing the logs.

Reply:
Why we are logging it also on lines 68 and 197? Should that be removed? I do not think that logging the error and returning it from the function is a good approach.

Reply:
Let me try to explain:
1. `trackChunks` should not have an effect on the smoke test - if it fails, we care about it, this is why we log it, but it doesn't return exit code in a way that marks the test as a failure.
2. We use `trackChunks` at two places
- before `fetch` so that we see where chunks are synced before retrieval, and

Reply:
Furthermore we care where an error happened within `trackChunks`. `getAllRefs` is an internal API, so should almost never fail, as it is in-process. The RPC calls however can fail for a variety of reasons and fail more often.

Bottom line - you are right that we might log the same error twice, but we don't really care about that level of detail, as long as we have enough information in the logs to determine what happened.

Feel free to propose an alternative solution if I am missing something.

Reply:
> Bottom line - you are right that we might log the same error twice, but we don't really care about that level of detail, as long as we have enough information in the logs to determine what happened.

I would be wrong if I say that I do not care about logging the same error twice. This is something that I consider a bad practice. If that is needed, then there is something wrong with our error propagation. I do not think that skipping on this principle contributes to the code cleanup and better architecture that we want to do.

Logging is very important, but it also should be correct, as someone can question why the same error happened twice during debugging. Or why the error needs double logging. If you really feel that all existing log lines are required, then that needs to be explained with comments to avoid that someone removes some of them in the future, breaking your expectations.

Reply:
removed the log line.

Reply:
Anton, please, do not think that I want to make you do something like "remove double logging, to make it more difficult to find where this error is coming from" in your latest commit https://github.com/ethersphere/go-ethereum/pull/1337/commits/476c30cf2666e2efddb2ce12a9991e259b1e7891.

I also want to have cleaner code and for that we should agree on some principles. My opinion is that the function caller should decide if it should log the error or pass it annotated to its caller. But, not both as it may result in double logging which can be confusing. By skipping on this types of things we will just end up with more confusion in the code. That is why I asked at least for the comment.

I would also like to opinions from others @frncmx @justelad. If the general understanding is that this is a non-issue I am happy to accept it.

Reply:
@janos you are missing the forest for the trees. We have more than 10 comments about a single log line in the `swarm-smoke` binary, the output of which only developers read. We are spending way too much time on the wrong thing.

The lifetime of this code will probably be short, since we will want to add much more functionality once we get to the point where the actual bugs in the `swarm` are addressed. Whether we log something once or twice in this test doesn't matter so much.

I really don't want to continue to have a discussion about a single log line, or whether this function has the best possible design - it is serving its purpose for the time being, and if it stops doing that in the near future - we will remove it and write a new one. This is most probably what's going to happen anyway, because tracing syncing chunk by chunk scales only so much - once we start running tests with bigger files, this won't be feasible in its current form.

Reply:
I'd much rather we focus on why we have hundreds of lines of dead code and unit tests along them in the codebase, rather than whether we log an error once or twice.